docs/architecture/engine.md

```diff
@@ -0,0 +1,270 @@
+   │     │    │  │ └───────┤   Blocks   ├────────┘  │   │     │      │
+   │     │    │  │         └────────────┘           │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │  ┌────────────────────────────┐  │   │     │      │
+   │     │    │  │  │                            │  │   │     │      │
+   │     │    │  │  │     Block [6PM - 8PM]      │  │   │     │      │
+   │     │    │  │  │                            │  │   │     │      │
+   │     │    │  │  ├────────────────────────────┤  │   │     │      │
+   │     │    │  │  │ Active Buffers (encoders)  │  │   │     │      │
+   │     │    │  │  └────────────────────────────┘  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │                                  │   │     │      │
+   │     │    │  │          ┌───────────┐           │   │     │      │
+   │     │    │  └──────────┤ Series 1  ├───────────┘   │     │      │
+   │     │    │             └───────────┘               │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │                                         │     │      │
+   │     │    │               ┌───────┐                 │     │      │
+   │     │    └───────────────┤Shard 1├─────────────────┘     │      │
+   │     │                    └───────┘                       │      │
+   │     │             ┌──────────────────────┐               │      │
+   │     └─────────────┤     Namespace 1      ├───────────────┘      │
+   │                   └──────────────────────┘                      │
+   │                                                                 │
+   │                                                                 │
+   │                                                                 │
+   │              ┌───────────────────────────────┐                  │
+   └──────────────┤           Database            ├──────────────────┘
+                  └───────────────────────────────┘
+```
+
+The in-memory portion of M3DB is implemented via a hierarchy of datastructures:
+
+1. A "database" (one per M3DB node, effectively a singleton)
```

Comment:
few nits:
- you're mixing case "database" and "Namespaces", could you pick one
- avoid using the word "singleton", too many negative connotations 
- "one per M3DB node" is vague - clarify what you mean - s/node/host|process

Reply:
good call, gonna standardize around lower case and will address the other two bullet points